torch/csrc/lazy/core/tensor.cpp

```diff
@@ -503,10 +503,12 @@ LazyTensorPtr GetOrCreateLtcTensor(const c10::optional<at::Tensor>& tensor,
 LazyTensorPtr GetLtcTensorOrCreateForWrappedNumber(const at::Tensor& tensor, const BackendDevice& device) {
   // TODO: There are places in core where a scalar is wrapped but not marked as
   // wrapped.
-  return (tensor.unsafeGetTensorImpl()->is_wrapped_number() ||
-          (tensor.dim() == 0 && tensor.numel() == 1))
-             ? GetOrCreateLtcTensor(tensor, device)
-             : GetLtcTensor(tensor);
+  return GetOrCreateLtcTensor(tensor, device);
```

Comment:
hmm, this is maybe a way to do it, but i wonder if we can leave the lazy tensor code alone and somehow cause the context manager to actually call .to(lazy) on the cuda tensors on __enter__ and likewise back to .to(cuda) with on __exit__?  We would need to do this in a way that did copyless transfers, of course.

Reply:
I remember this proposal in the doc.

> actually call .to(lazy) on the cuda tensors on enter and likewise back to .to(cuda) with on exit?

^^^ this is the part where I'm concerned how it could be done efficiently.
e.g. when we exit the context, we want to call `.to(cuda)` on all lazy tensors. Are we going to just rely on ref-count to figure out what tensors would be live after exiting the context? Would that be too aggressive of a materialization strategy?

Reply:
I guess the entrance problem will be even trickier... e.g. if lazy context only encloses a small chunk of the a large model with many live eager tensors, how do we predict what tensors would be used within the context and should be packaged as lazy tensors?

Reply:
I think the entrance problem would be solved 'lazily'- we would not have a point in time where we try to bulk-convert a set of eager tensor into lazy tensors.  Instead, we would just allow eager tensors as input to lazy computations; when this happened, the lazy kernel would have to gracefully wrap/promote those tensors (.to(lazy)) in an efficient way (no copy on the underlying cuda data).  This also means we'd have to have some way to tell the internals of lazy to behave differently under this mode, vs as a device, for compatibility with XLA backend.

The exit would have to account for 2 things
-(a) modifications to 'input' eager tensors
-(b) allowing 'new' lazy tensors to be returned back to eager
For (a), presumably we could just rely on the fact that we're zero-copying the input data buffer and then push an update back to that same buffer as we run the lazy computation.  For (b), what would we do though? If the python variable bound to a lazy tensor is still reachable after the mode-scope-exit, then we could probably just make it so accessing it outside the mode scope silently converted it to the eager tensor that its holding inside as its 'tensor_data'.